listings/applications/simple_vault/src/simple_vault.cairo

```diff
@@ -91,11 +116,114 @@ pub mod SimpleVault {
+        contract_address_const
+    };
+
+    const token_name: felt252 = 'myToken';
+    const decimals: u8 = 18;
+    const initial_supply: felt252 = 100000;
+    const symbols: felt252 = 'mtk';
+
+    fn deploy() -> (ISimpleVaultDispatcher, ContractAddress, IERC20Dispatcher_token) {
+        let _token_address: ContractAddress = contract_address_const::<'token_address'>();
+        let caller = contract_address_const::<'caller'>();
+
+        let (token_contract_address, _) = deploy_syscall(
+            erc20::token::erc20::TEST_CLASS_HASH.try_into().unwrap(),
+            caller.into(),
+            array![caller.into(), 'myToken', '8', '1000'.into(), 'MYT'].span(),
+            false
+        )
+            .unwrap_syscall();
+
+        let (contract_address, _) = deploy_syscall(
+            SimpleVault::TEST_CLASS_HASH.try_into().unwrap(),
+            0,
+            array![token_contract_address.into()].span(),
+            false
+        )
+            .unwrap_syscall();
 
+        (
+            ISimpleVaultDispatcher { contract_address },
+            contract_address,
+            IERC20Dispatcher_token { contract_address: token_contract_address }
+        )
+    }
+
+    #[test]
+    fn test_deposit() {
+        let caller = contract_address_const::<'caller'>();
+        let (dispatcher, vault_address, token_dispatcher) = deploy();
+
+        // Approve the vault to transfer tokens on behalf of the caller
+        let amount: felt252 = 10.into();
+        token_dispatcher.approve(vault_address.into(), amount);
+        set_contract_address(caller);
+
+        // Deposit tokens into the vault
+        let amount: u256 = 10.into();
+        let _deposit = dispatcher.deposit(amount);
+        println!("deposit :{:?}", _deposit);
+
+        // Check balances and total supply
+        let balance_of_caller = dispatcher.user_balance_of(caller);
+        let total_supply = dispatcher.contract_total_supply();
+
+        assert(balance_of_caller == amount, 'Deposit failed');
+        assert(total_supply == amount, 'total supply mismatch');
+
+    }
+
+    #[test]
+    fn test_deposit_withdraw() {
+        let caller = contract_address_const::<'caller'>();
+        let (dispatcher, vault_address, token_dispatcher) = deploy();
+
+        // Approve the vault to transfer tokens on behalf of the caller
+        let amount: felt252 = 10.into();
+        token_dispatcher.approve(vault_address.into(), amount);
+        set_contract_address(caller);
+        set_account_contract_address(vault_address);
+
+        // Deposit tokens into the vault
+        let amount: u256 = 10.into();
+        dispatcher.deposit(amount);
+        dispatcher.withdraw(amount);
+
+        // Check balances of user in the vault after withdraw
+        let balance_of_caller = dispatcher.user_balance_of(caller);
+
+        assert(balance_of_caller == 0.into(), 'withdraw failed');
+    }
+   
```

Comment:
Could you try to create a test for the yield feature? If you find this too complex, it's ok for now but it would be a nice addition!

Reply:
I have tried to simulate yield feature in a test but it's failing 




```cairo
#[test]
fn test_yield_distribution() {
    let caller = contract_address_const::<'caller'>();
    let (dispatcher, vault_address, token_dispatcher) = deploy();

    // Initial setup
    let initial_deposit: felt252 = 10.into();
    token_dispatcher.approve(vault_address.into(), initial_deposit);
    set_contract_address(caller);

    // Deposit tokens into the vault
    let deposit_amount: u256 = 10.into();
    dispatcher.deposit(deposit_amount);

    // Simulate yield generation by transferring tokens directly to the vault
    let yield_amount: felt252 = 5.into();
    let vault_contract_address: ContractAddress = vault_address.into();
    token_dispatcher.transfer(vault_contract_address, yield_amount); <---- the 5 token did not enter the vault contract.

    // Check balances and total supply after yield
    let expected_total_supply: u256 = 10.into(); // Total supply should remain the same
    let total_supply = dispatcher.contract_total_supply();
    assert(total_supply == expected_total_supply, "Total supply mismatch after yield");

    // Check user's new balance
    let user_balance = dispatcher.user_balance_of(caller);
    let expected_user_balance: u256 = 15.into(); // Initial deposit + yield
    assert(user_balance == expected_user_balance, "Yield distribution failed"); <------ its failing here because the 5 token transfered to the vault does not reflect on the vault contract
}




```

Reply:
@julio4, I am still waiting for your response on this 

> I have tried to simulate yield feature in a test but it's failing
> 
> ```cairo
> #[test]
> fn test_yield_distribution() {
>     let caller = contract_address_const::<'caller'>();
>     let (dispatcher, vault_address, token_dispatcher) = deploy();
> 
>     // Initial setup
>     let initial_deposit: felt252 = 10.into();
>     token_dispatcher.approve(vault_address.into(), initial_deposit);
>     set_contract_address(caller);
> 
>     // Deposit tokens into the vault
>     let deposit_amount: u256 = 10.into();
>     dispatcher.deposit(deposit_amount);
> 
>     // Simulate yield generation by transferring tokens directly to the vault
>     let yield_amount: felt252 = 5.into();
>     let vault_contract_address: ContractAddress = vault_address.into();
>     token_dispatcher.transfer(vault_contract_address, yield_amount); <---- the 5 token did not enter the vault contract.
> 
>     // Check balances and total supply after yield
>     let expected_total_supply: u256 = 10.into(); // Total supply should remain the same
>     let total_supply = dispatcher.contract_total_supply();
>     assert(total_supply == expected_total_supply, "Total supply mismatch after yield");
> 
>     // Check user's new balance
>     let user_balance = dispatcher.user_balance_of(caller);
>     let expected_user_balance: u256 = 15.into(); // Initial deposit + yield
>     assert(user_balance == expected_user_balance, "Yield distribution failed"); <------ its failing here because the 5 token transfered to the vault does not reflect on the vault contract
> }
> ```